lib/graphql/execution/instrumentation.rb

```diff
@@ -0,0 +1,82 @@
+# frozen_string_literal: true
+module GraphQL
+  module Execution
+    module Instrumentation
+      # This function implements the instrumentation policy:
+      #
+      # - Instrumenters are a stack; the first `before_query` will have the last `after_query`
+      # - If a `before_` hook returned without an error, its corresponding `after_` hook will run.
+      # - If the `before_` hook did _not_ run, the `after_` hook will not be called.
+      #
+      # When errors are raised from `after_` hooks:
+      #   - Subsequent `after_` hooks _are_ called
+      #   - The first raised error is captured; later errors are ignored
+      #   - If an error was capture, it's re-raised after all hooks are finished
+      #
+      # Partial runs of instrumentation are possible:
+      # - If a `before_multiplex` hook raises an error, no `before_query` hooks will run
+      # - If a `before_query` hook raises an error, subsequent `before_query` hooks will not run (on any query)
+      def self.apply_instrumenters(multiplex)
+        schema = multiplex.schema
+        queries = multiplex.queries
+        query_instrumenters = schema.instrumenters[:query]
+        multiplex_instrumenters = schema.instrumenters[:multiplex]
+
+        # First, run multiplex instrumentation, then query instrumentation for each query
+        call_hooks(multiplex_instrumenters, multiplex, :before_multiplex, :after_multiplex) do
+          each_query_call_hooks(query_instrumenters, queries) do
+            # Let them be executed
+            yield
+          end
+        end
+      end
+
+      class << self
+        private
+        # Call the before_ hooks of each query,
+        # Then yield if no errors.
+        # `call_hooks` takes care of appropriate cleanup.
+        def each_query_call_hooks(instrumenters, queries, i = 0)
+          if i >= queries.length
+            yield
+          else
+            query = queries[i]
+            call_hooks(instrumenters, query, :before_query, :after_query) {
+              each_query_call_hooks(instrumenters, queries, i + 1) {
```

Comment:
I didn't think about the fact that you were calling the query hooks for all queries before and after all queries, instead of around each query.  I guess that is because lazy execution is meant to overlap across all the queries.  Do you expect a large number of queries would often get multiplexed together?  Since this would increase the stack size by O(Q * I) where `Q` is the number of queries and `I` is the number of query instrumenters.

If this is a concern or does become a problem, then I thought of a more complicated solution could be used to collapse the stack.

```ruby
def call_after_hooks(instrumenters, object, after_hook_name, i)
  while i >= 0
    instrumenter = instrumenters.fetch(i)
    i -= 1
    instrumenter.public_send(after_hook_name, object)
  end
ensure
  # Call remaining hooks even if an exception or `throw` is used.
  # Exception#cause can be used to access all exceptions when multiple
  # after hooks raise an exception
  if i >= 0
    call_after_hooks(instrumenters, object, after_hook_name, i)
  end
end

def call_hooks(instrumenters, object, before_hook_name, after_hook_name, i=0)
  instrumenters.each do |instrumenter|
    instrumenter.public_send(before_hook_name, object)
    i += 1
  end
  yield
ensure
  call_after_hooks(instrumenters, object, after_hook_name, i - 1)
end
```

We don't multiplex multiple queries, so what you have right now works for us.

Reply:
Cool, I'll tuck this away and cross that bridge when we come to it!